File: 2029-stone-game-ix/2029-stone-game-ix.cpp
class Solution {
public:
    bool stoneGameIX(vector<int>& stones) {
        // Count frequencies of stone values modulo 3
        vector<int> count(3, 0);
        for (int stone : stones) {
            count[stone % 3]++;
        }
        
        // Case 1: The number of stones with remainder 0 is even
        if (count[0] % 2 == 0) {
            // Alice needs at least one of each to play optimally and win
            return min(count[1], count[2]) > 0;
        }
        
        // Case 2: The number of stones with remainder 0 is odd
        // Alice needs a significant difference to force Bob into losing
        return abs(count[1] - count[2]) > 2;
    }
};

File: 2029-stone-game-ix/2029-stone-game-ix.java
class Solution {
    public boolean stoneGameIX(int[] stones) {
        // Count frequencies of stones based on their remainders when divided by 3
        int[] count = new int[3];
        for (int stone : stones) {
            count[stone % 3]++;
        }
        
        // Case 1: The number of remainder 0 stones is even
        if (count[0] % 2 == 0) {
            // Alice can win if there is at least one stone of type 1 and one stone of type 2
            return Math.min(count[1], count[2]) > 0;
        }
        
        // Case 2: The number of remainder 0 stones is odd
        // Alice can win if one side has more than 2 extra stones to shift the final turn penalty
        return Math.abs(count[1] - count[2]) > 2;
    }
}